benchmark/benchmark.go

```diff
@@ -0,0 +1,258 @@
+		FROM repositories`,
+	},
+	{
+		"Last commit messages in HEAD for every repository",
+		`
+		SELECT c.message
+		FROM
+			refs r
+			JOIN commits c ON r.hash = c.hash
+		WHERE
+			r.name = 'refs/heads/HEAD';`,
+	},
+	{
+		"All commit messages in HEAD history for every repository",
+		`
+		SELECT c.message
+		FROM
+			commits c
+			JOIN refs r ON r.hash = c.hash
+		WHERE
+			r.name = 'refs/heads/HEAD' AND
+			history_idx(r.hash, c.hash) >= 0;`,
+	},
+	{
+		"Top 10 repositories by commit count in HEAD",
+		`
+		SELECT
+			repository_id,
+			commit_count
+		FROM (
+			SELECT
+				r.repository_id,
+				count(*) AS commit_count
+			FROM
+				refs r
+				JOIN commits c ON history_idx(r.hash, c.hash) >= 0
+			WHERE
+				r.name = 'refs/heads/HEAD'
+			GROUP BY r.repository_id
+		) AS q
+		ORDER BY commit_count DESC
+		LIMIT 10;`,
+	},
+	{
+		"Count repository HEADs",
+		`
+		SELECT
+			COUNT(DISTINCT r.repository_id) AS head_count
+		FROM
+			refs r
+		WHERE name = 'refs/heads/HEAD';`,
+	},
+	{
+		"Repository count by language presence (HEAD, no forks)",
+		`
+		SELECT *
+		FROM (
+			SELECT
+				language,
+				COUNT(repository_id) AS repository_count
+			FROM (
+				SELECT DISTINCT
+					r.repository_id AS repository_id,
+					language(t.name, b.content) AS language
+				FROM
+					refs r
+					JOIN commits c ON r.hash = c.hash
+					JOIN tree_entries t ON commit_has_tree(c.hash, t.tree_hash)
+					JOIN blobs b ON t.entry_hash = b.hash
+				WHERE
+					r.name = 'refs/heads/HEAD'
+			) AS q1
+			GROUP BY language
+		) AS q2
+		ORDER BY repository_count DESC;`,
+	},
+	{
+		"Top 10 repositories by contributor count (all branches)",
+		`
+		SELECT
+			repository_id,
+			contributor_count
+		FROM (
+			SELECT
+				repository_id,
+				COUNT(DISTINCT c.author_email) AS contributor_count
+			FROM
+				refs r
+				JOIN commits c ON history_idx(r.hash, c.hash) >= 0
+			GROUP BY repository_id
+		) AS q
+		ORDER BY contributor_count DESC
+		LIMIT 10;`,
+	},
+	{
+		"Created projects per year",
+		`
+		SELECT
+			year,
+			COUNT(DISTINCT hash) AS project_count
+		FROM (
+			SELECT
+				hash,
+				YEAR(author_date) AS year
+			FROM
+				refs r
+				JOIN commits c ON r.hash = c.hash
+			WHERE
+				r.name = 'refs/heads/HEAD'
+		) AS q
+		GROUP BY year
+		ORDER BY year DESC;`,
+	},
+}
+
+func main() {
+	path, err := getDatasetPath()
+	if err != nil {
+		fatal("[ERROR] unable to get dataset path: %s", err)
+	}
+
+	pool, err := newPool(path)
+	if err != nil {
+		fatal("[ERROR] unable to create repository pool: %s", err)
+	}
+
+	engine := newEngine()
+
+	suite.run(pool, engine)
+}
+
+type benchmark struct {
+	name  string
+	query string
+}
+
+func (b benchmark) run(pool *gitbase.RepositoryPool, e *sqle.Engine) error {
+	info("[RUN] %s", b.name)
+	info(b.query)
+	session := gitbase.NewSession(pool)
+	ctx := sql.NewContext(context.TODO(), sql.WithSession(session))
+	start := time.Now()
+
+	_, iter, err := e.Query(ctx, string(b.query))
```

Comment:
Why is this casting necessary? Isn't `b.query` already a string?

Reply:
Leftover, before the benchmark type was an aliased string